src/workspacer.Sound/SoundManager.cs

```diff
@@ -0,0 +1,393 @@
+            }
+
+            foreach (var client in clients)
+            {
+                action(client);
+            }
+        }
+
+        #region System Events
+
+        public void OnDefaultDeviceChanged(EDataFlow type, ERole role, string defaultDeviceId)
+        {
+            ApplyToClientsWithCriteria(string.Empty, type.ConvertToDeviceType(), role.ConvertToDeviceRole(), (client) =>
+            {
+                _mmDeviceEnumerator.GetDevice(defaultDeviceId, out var device);
+                if (device == null)
+                {
+                    return;
+                }
+
+                client.SetDevice(device);
+                client.Refresh();
+            });
+
+#if EVENT_DEBUG
+            MessageBox.Show($"DeviceRemoved DataFlow:[{dataFlow}] DeviceRole:[{deviceRole}] DeviceId:[{defaultDeviceId}]");
+#endif
+        }
+
+        public void OnDeviceStateChanged(string deviceId, uint newState)
+        {
+            ApplyToClientsWithCriteria(deviceId, null, null, (client) => client.Refresh() );
+
+            DeviceChanged?.Invoke(new DeviceInfo { Id = deviceId }, newState);
+#if EVENT_DEBUG
+            MessageBox.Show($"DeviceStateChanged deviceId:[{deviceId}] newState:[{newState}]");
+#endif
+        }
+
+        public void OnDeviceAdded(string deviceId)
+        {
+            ApplyToClientsWithCriteria(deviceId, null, null, (client) => client.Refresh());
+
+            DeviceAdded?.Invoke(new DeviceInfo { Id = deviceId });
+#if EVENT_DEBUG
+            MessageBox.Show($"DeviceAdded DeviceId:[{deviceId}]");
+#endif
+        }
+
+        public void OnDeviceRemoved(string deviceId)
+        {
+            ApplyToClientsWithCriteria(deviceId, null, null, (client) => client.Refresh());
+
+            DeviceRemoved?.Invoke(new DeviceInfo { Id = deviceId });
+#if EVENT_DEBUG
+            MessageBox.Show($"DeviceRemoved DeviceId:[{deviceId}]");
+#endif
+        }
+
+        public void OnPropertyValueChanged(string deviceId, PROPERTYKEY propertyKey)
+        {
+            ApplyToClientsWithCriteria(deviceId, null, null, (client) => client.Refresh());
+
+            DevicePropertyValueChanged?.Invoke(new DeviceInfo { Id = deviceId }, propertyKey);
+#if EVENT_DEBUG
+            MessageBox.Show($"DeviceStateChanged deviceId:[{deviceId}] propertyKey:[{propertyKey}]");
+#endif
+        }
+        #endregion
+    }
+
+    public class SoundClient : IAudioEndpointVolumeCallback
+    {
+        public event DeviceDataEventDelegate<float> VolumeChanged;
+        public event DeviceDataEventDelegate<bool> MuteChanged;
+
+        private IMMDevice _device;
+
+        public DeviceInfo DeviceInfo { get; set; }
+
+        private IAudioEndpointVolume _deviceAudioEndpointVolume;
+
+        public void SetDevice(IMMDevice device)
+        {
+            _device = device;
+            InitializeAudioEnpoint(device);
+        }
+
+        private void InitializeAudioEnpoint(IMMDevice device = null)
```

Comment:
Do we need this argument to be optional? Do we need it at all? When would be initialize another audio-endpoint than the one the object represents?